File: browser/src/slideshow/SlideCache.ts
// @ts-strict-ignore
/** */

/*
 * Copyright the Collabora Online contributors.
 *
 * SPDX-License-Identifier: MPL-2.0
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * SlideCache is storing the cached layers
 */

declare var SlideShow: any;

class SlideCache {
	private renderedSlides: Map<string, ImageBitmap> = new Map();
	// TODO: cache other layers here

	public has(slideHash: string) {
		return this.renderedSlides.has(slideHash);
	}

	public get(slideHash: string): ImageBitmap {
		return this.renderedSlides.get(slideHash);
	}

	public set(slideHash: string, image: ImageBitmap): void {
		this.renderedSlides.set(slideHash, image);
	}

	public invalidate(slideHash: string): void {
		this.renderedSlides.delete(slideHash);
	}

	public invalidateAll(): void {
		this.renderedSlides.clear();
	}
}

SlideShow.SlideCache = SlideCache;

File: wsd/SlideCache.cpp
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * Copyright the Collabora Online contributors.
 *
 * SPDX-License-Identifier: MPL-2.0
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include "SlideCache.hpp"

void SlideLayerCacheMap::insert(const std::string& key, const std::shared_ptr<Message> cachedData)
{
    insertion_order.push(key);
    cache_map[key].push_back(cachedData);
}

std::size_t SlideLayerCacheMap::reduceSizeTo(std::size_t desiredSize)
{
    if (cache_map.size() <= desiredSize)
        return 0;

    std::size_t total_deleted_entries = 0;
    while (cache_map.size() > desiredSize)
    {
        cache_map.erase(insertion_order.front());
        insertion_order.pop();
        total_deleted_entries++;
    }

    return total_deleted_entries;
}

void SlideLayerCacheMap::erase_all()
{
    cache_map.clear();
    std::queue<std::string> empty;
    std::swap(insertion_order, empty);
}

std::unordered_map<std::string, std::vector<std::shared_ptr<Message>>>::const_iterator
SlideLayerCacheMap::find(const std::string& key) const
{
    return cache_map.find(key);
}

std::unordered_map<std::string, std::vector<std::shared_ptr<Message>>>::const_iterator
SlideLayerCacheMap::end()
{
    return cache_map.end();
}
